service/storage_service.cc

```diff
@@ -6033,8 +6146,26 @@ future<> storage_service::update_topology_change_info(mutable_token_metadata_ptr
     assert(this_shard_id() == 0);
 
     try {
-        locator::dc_rack_fn get_dc_rack_from_gossiper([this] (inet_address ep) { return get_dc_rack_for(ep); });
-        co_await tmptr->update_topology_change_info(get_dc_rack_from_gossiper);
+        locator::dc_rack_fn get_dc_rack_by_host_id([this, &tm = *tmptr] (locator::host_id host_id) -> std::optional<locator::endpoint_dc_rack> {
+            if (_raft_topology_change_enabled) {
+                const auto server_id = raft::server_id(host_id.uuid());
+                const auto* node = _topology_state_machine._topology.find(server_id);
+                if (node) {
+                    return locator::endpoint_dc_rack {
+                        .dc = node->second.datacenter,
+                        .rack = node->second.rack,
+                    };
+                }
+                return std::nullopt;
+            }
+
+            const auto ep = tm.get_endpoint_for_host_id_if_known(host_id);
```

Comment:
Why don't we map this through gossiper here? Looks like gossiper should be the source of truth when we're calling `get_dc_rack_by_host_id`.

Or is it guaranteed that by the moment this function is called, token metadata has already been updated with mappings from gossiper? (If so, perhaps we should comment on that precondition)

Reply:
gossiper has method `get_nodes_with_host_id`, which may return many endpoints for a host_id. We rely on token_metadata.topology to represent the current actual mapping, we strive to maintain this assumption throughout gossiper-managed topology transitions.

Reply:
> We rely on token_metadata.topology to represent the current actual mapping, we strive to maintain this assumption throughout gossiper-managed topology transitions.

So we assume that when this lambda is called, `tm` already has the current mapping, and not e.g. is in the middle of being updated from gossiper as well.

Reply:
> We rely on token_metadata invariant, that its topology should contain IPs for all known host_id-s.

For a certain definition of "known". There is a point where host ID transitions from unknown to known, i.e. we take information about it from gossiper and put into token_metadata (together with its IP -- that's the invariant).

The question is, is it guaranteed that this lambda is called *after* the point of inserting host ID (together with its IP) into token_metadata, or could it be called before -- for example, precisely as part of calculation needed to do the token_metadata insertion?

Reply:
> I'll try to change this to a more strict get_endpoint_for_host_id, this should work based on this logic.

It's kinda weird now that the code now allows dc/rack to be missing on raft-topology path, but not gossiper path

Reply:
Well ok `get_dc_rack_for` may return nullopt still.

I wonder which scenarios this could happen on though.